File: app/src/main/java/com/mklc/endlessrecyclerapp/ui/main/UserViewModel.java
package com.mklc.endlessrecyclerapp.ui.main;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.mklc.endlessrecyclerapp.data.model.network.User;
import com.mklc.endlessrecyclerapp.data.network.repository.UserRepository;

import java.util.List;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class UserViewModel extends ViewModel {

    private final UserRepository userRepository;

    @Inject
    public UserViewModel(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    private MutableLiveData<List<User>> liveDataUserList;

    public MutableLiveData<List<User>> getLiveDataUserList() {
        if (liveDataUserList == null) {
            liveDataUserList = new MutableLiveData<>();
        }
        return liveDataUserList;
    }

    public void loadUsers(int page) {

        getLiveDataUserList().setValue(null);

        userRepository.loadUsers(page).enqueue(new Callback<List<User>>() {
            @Override
            public void onResponse(@NonNull Call<List<User>> call, @NonNull Response<List<User>> response) {
                if (response.isSuccessful()) {
                    if (response.body() != null) {
                        getLiveDataUserList().setValue(response.body());
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<List<User>> call, @NonNull Throwable t) {
                getLiveDataUserList().postValue(null);
            }
        });
    }
}

File: app/src/main/java/com/mklc/endlessrecyclerapp/ui/main/UserViewModel.kt
package com.mklc.endlessrecyclerapp.ui.main

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.mklc.endlessrecyclerapp.data.UserDto
import com.mklc.remote.data.model.response.UserResponse
import com.mklc.remote.repository.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import timber.log.Timber
import javax.inject.Inject

@HiltViewModel
class UserViewModel @Inject constructor(
    private val userRepository: UserRepository,
) : ViewModel() {

    private val _isLoading = MutableLiveData(false)

    private val userList = mutableListOf<UserDto>()
    val liveDataUserList = MutableLiveData<List<UserDto>>()
    val liveDataStopListLoading = MutableLiveData(false)

    fun loadUsers(page: Int){
        if(page > 4){ // check if it will really stop
            stopLoading()
            return
        }
        userRepository.loadUsers(page).subscribe({
           setupUsers(it)
        },{
            stopLoading()
            Timber.e(it)
        })
    }

    private fun setupUsers(response: List<UserResponse>){
        userList.addAll(response.map { user -> UserDto(user.id, user.name, user.email, user.status) })

        liveDataUserList.postValue(userList.toList())
    }

    fun isLoading() = _isLoading

    private fun stopLoading(){
        liveDataStopListLoading.postValue(true)
    }
}